### DISCUSSION TITLE
Flushed content may disappear when another session is closed.

### DISCUSSION THREAD

**Discussion by irreg** [sqlalchemy/sqlalchemy#8643] [Usage Questions]
### Describe the bug

After flushing one session, perform a read-only operation in another session, and when closed, the flushed contents disappear.

### To Reproduce

```python
from sqlalchemy import create_engine
from sqlalchemy import Column, Integer
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker

Base = declarative_base()


class Test(Base):
    __tablename__ = 'test'

    id = Column(Integer, primary_key=True)

engine = create_engine('sqlite:///:memory:')
Base.metadata.create_all(engine)
Session = sessionmaker(bind=engine)
session1 = Session()
session2 = Session()
session2.query(Test).all()

session1.add(Test())
session1.flush()

session2.close()

print(session1.query(Test).all()) # Cannot retrieve flushed value
```

### Error

```
# Copy complete stack trace and error message here, including SQL log output if applicable.
``` 

### Versions

- OS: Windows 10 Pro 212H/11 Pro
- Python: 3.8.5/3.8.9
- SQLAlchemy: 1.4.35/1.4.41
- Database: SQLite
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): default(built-in sqlite3)

**Comment by CaselIT** [comment]
Hi,

Using sqlite memory with multiple sessions will share the same connection so the rollback will rollback the shared connection.

Try using a file with sqlite or another db

**Reply by irreg** [reply]
Thanks, I understand the behavior.

We are using SQLCipher (sqlcipher3) in the production environment.
This seems to cause the same behavior because the default is SingletonThreadPool. (In my personal opinion, I feel that it would be better to change the value set by default)

I will try to improve this by changing to QueuePool (SQLCipher has a high cost of opening connections, so other pools don't seem to work well...).

**Reply by zzzeek** [reply]
QueuePool is not compatible with a plain ":memory:" database.  each new connection would use a brand new database.  surely that's not what you wanted?

**Reply by irreg** [reply]
I didn't say enough.
My goal is to implement the process using SQLCipher.
I originally faced this behavior when trying to use SQLCipher, and since I had the same behavior with SQLite(:memory:), I thought something was wrong.
I understand that QueuePool does not work in ":memory:", I will make the Pool change only for the DB of SQLCipher.
Thanks.
